### ISSUE TITLE
Figure subcaptions are randomly arranged in an order I didn't set in a python chunk in a qmd file.

### ISSUE THREAD

**Issue by kmink3225** [quarto-dev/quarto-cli#5508]
### Bug description

Hello, To briefly explain the situation, I am an R user, but since all of the project participants are python users, I was conducting statistical analysis using both python and R. However, when using R chunk and python chunk at the same time, it seems that the order of subcaptions is not rendered as I intended. Should the simultaneous use of multiple quarto engines be discouraged? In other words, if I choose to use Python, should I stick with Python?

````
---
title-block-banner: true
title: "Something1"
subtitle: "Something2"
date: 2023-04-03
author: "Someone"
institute: "Something3"
format:
  html:
  #pdf:
    #documentclass: scrartcl
    #papersize: a4
    toc: true
    number-sections: true
    lot: true
    lof: true
    code-fold: true
    #df-print: paged
    fig-width: 15
    fig-height: 15
    #fontfamily: libertinus
    #colorlinks: true
execute:
  warning: false
  message: false
  echo: false
---

# Analysis With R 

## Load Packages

```{r}
#| eval: false

# install.packages('tidyverse')
# install.packages('reticulate')
# install.packages('arrow')

library(tidyverse)
library(reticulate)
library(knitr)
library(kableExtra)
#library(arrow)
#library(readr)
options(dplyr.summarise.inform = FALSE)
```


# Anlaysis with Python

## Load Packages

```{python load_packages}
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn 
import itertools
import os
from IPython.display import display, Markdown # for dynamic documentation
from tabulate import tabulate # for dynamic documentation
from pandas.api.types import CategoricalDtype # for factoring categorical variables
from sklearn.metrics import confusion_matrix #, accuracy_score, precision_score, recall_score, f1_score

pd.set_option('display.max_rows', 100)
# pd.set_option('display.max_columns', 10)
# pd.set_option('display.width', 1000)
pd.options.mode.chained_assignment = None 
```

## Data Load

```{python}
# Long form data generation
data1 = {'dataid': np.repeat(range(0,100), 6),
        'cycle': np.tile(range(6), 100),
        'rfu': np.random.randn(600),
        'scenario': np.repeat('scenario00',600),
        'template': np.repeat([chr(i) for i in range(65, 71)],100)}

# Convert to pandas DataFrame
unnested_dsp_error_data = pd.DataFrame(data1)

```

## Function List

```{python, funtions}

def plot_error_signals(dataid_list,title=None,x_limits=None,y_limits=None):
  temp2=unnested_dsp_error_data[unnested_dsp_error_data.dataid.isin(dataid_list)]
  plt.title(title)
  for i in dataid_list:
    temp=unnested_dsp_error_data[(unnested_dsp_error_data.dataid==i)]
    plt.plot(temp.cycle,temp.rfu)
    if x_limits or y_limits:
      plt.ylim(x_limits,y_limits)
    else:
      plt.ylim(temp2.rfu.min()*0.9,temp2.rfu.max()*1.1)

```

## Qualititative Analysis

```{python, DSP_grand_confusion_errors}
#| label: dsp_grand_confusion_errors
#| fig-cap: "DSP Grand Missclassifed Signals"
#| echo: fenced

error_dataid=unnested_dsp_error_data.dataid.unique().tolist()

plot_error_signals(error_dataid)
plt.show()
```

```{python, DSP_error_signals_templates}
#| label: DSP_error_signals_templates
#| fig-cap: "DSP Error Signals Grouped By Templates"
#| fig-subcap: 
#|   - "A"
#|   - "B"
#|   - "C"
#|   - "D"
#|   - "E"
#|   - "F"
#| layout-ncol: 2
#| echo: fenced

template_list=[chr(i) for i in range(65, 71)]

for template in template_list:
  temp_dataid=unnested_dsp_error_data[unnested_dsp_error_data.template==template].dataid.unique()
  plot_error_signals(temp_dataid,title=template)
  plt.show()

```




````

the rendered output is like the following:

![image](https://github.com/quarto-dev/quarto-cli/assets/61118160/bd9e5d95-cdc1-441c-975b-199ffe3545d6)

### Checklist

- [X] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [x] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [x] Please document the operating system you're running. If on Linux, please provide the specific distribution.
- [x] Please provide the output of `quarto check` so we know which version of quarto and its dependencies you're running.

**Comment by mcanouil** [comment]
Again, we do really need the information from the checklist in order to reproduce.
For example your (not minimal) example relies on deprecated Python module, Python, R, and reticulate.
Using the latest of everything, I can't run your example without having to do extensive code checks and fixing/updating it.

Minimal means we don't need to have commented code/options, only what's really important to make the example work.
If you are running Python through R via reticulate, instead of adding a non evaluated code cell, simply add `engine: knitr` in your yaml.

**Comment by kmink3225** [comment]
I am sorry to miss what you asked. I did not understand the checklist when I reported the issue. But, I think I understood better reading your reply :). I tried to answer to the checklist. Please, let me know what I have to add for the reproduction. Just in case, I also remain for your reproduction the github repository address: https://github.com/kmink3225/quarto_error.git .

## minimal code

````
---
title-block-banner: true
title: "Something1"
subtitle: "Something2"
date: 2023-04-03
author: "Someone"
institute: "Something3"
format:
  html:
    code-fold: true
execute:
  warning: false
  message: false
  echo: false
---

# Analysis With R 

## Load Packages

```{r}
#| eval: false

library(tidyverse)
library(reticulate)
library(knitr)
library(kableExtra)
options(dplyr.summarise.inform = FALSE)
```

```{python}
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
#import seaborn as sns
#import sklearn 
#import itertools
#import os
#from IPython.display import display, Markdown # for dynamic documentation
#from tabulate import tabulate # for dynamic documentation
#from pandas.api.types import CategoricalDtype # for factoring categorical variables
#from sklearn.metrics import confusion_matrix #, accuracy_score, precision_score, recall_score, f1_score

#pd.set_option('display.max_rows', 100)
# pd.set_option('display.max_columns', 10)
# pd.set_option('display.width', 1000)
#pd.options.mode.chained_assignment = None 
```

## Data Load

```{python}
# Long form data generation
data1 = {'dataid': np.repeat(range(0,100), 6),
        'cycle': np.tile(range(6), 100),
        'rfu': np.random.randn(600),
        'template': np.repeat([chr(i) for i in range(65, 71)],100)}

# Convert to pandas DataFrame
unnested_dsp_error_data = pd.DataFrame(data1)

def plot_function(dataid_list,title=None,x_limits=None,y_limits=None):
  temp2=unnested_dsp_error_data[unnested_dsp_error_data.dataid.isin(dataid_list)]
  plt.title(title)
  for i in dataid_list:
    temp=unnested_dsp_error_data[(unnested_dsp_error_data.dataid==i)]
    plt.plot(temp.cycle,temp.rfu)
    if x_limits or y_limits:
      plt.ylim(x_limits,y_limits)
    else:
      plt.ylim(temp2.rfu.min()*0.9,temp2.rfu.max()*1.1)
```

```{python}
#| label: grand_signals
#| fig-cap: "Grand Signals"
#| echo: fenced

error_dataid=unnested_dsp_error_data.dataid.unique().tolist()

plot_function(error_dataid)
plt.show()
```

```{python}
#| label: grouped_signals
#| fig-cap: "Grouped Signals"
#| fig-subcap: 
#|   - "A"
#|   - "B"
#|   - "C"
#|   - "D"
#|   - "E"
#|   - "F"
#| layout-ncol: 2
#| echo: fenced

template_list=[chr(i) for i in range(65, 71)]

for template in template_list:
  temp_dataid=unnested_dsp_error_data[unnested_dsp_error_data.template==template].dataid.unique()
  plot_function(temp_dataid,title=template)
  plt.show()

```


````

## quarto check

````
A new release of Deno is available: 1.28.2 → 1.33.3 Run `deno upgrade` to install it.

[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.1: OK
      Dart Sass version 1.55.0: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.3.353
      Path: C:\Users\kmkim\AppData\Local\Programs\Quarto\bin
      CodePage: 949

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....OK
      Version: 3.9.16 (Conda)
      Path: C:/Users/kmkim/.conda/envs/study/python.exe
      Jupyter: 5.2.0
      Kernels: python3

[>] Checking Jupyter engine render....OK

[>] Checking R installation...........OK
      Version: 4.2.2
      Path: C:/R/R-4.2.2
      LibPaths:
        - C:/R/R-4.2.2/library
      knitr: 1.42
      rmarkdown: 2.21

[>] Checking Knitr engine render......OK
````

## OS and IDE

### OS

edition: Windows 10 Enterprise
version: 22H2
installed date: ‎2023-‎02-‎01
OS build:	19045.2965
Windows Feature Experience Pack 1000.19041.1000.0
system type: 64 -bit

### IDE
#### R Studio
RStudio 2023.03.1+446 "Cherry Blossom" Release (6e31ffc3ef2a1f81d377eeccab71ddc11cfbd29e, 2023-05-09) for windows
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) RStudio/2023.03.1+446 Chrome/108.0.5359.179 Electron/22.0.3 Safari/537.36

#### VS Code 
version: 1.78.2 (user setup)
Commit: b3e4e68a0bc097f0ae7907b217c1119af9e03435
Date: 2023-05-10T14:39:26.248Z
Electron: 22.5.2
Chromium: 108.0.5359.215
Node.js: 16.17.1
V8: 10.8.168.25-electron.0
OS: Windows_NT x64 10.0.19045
Sandbox: No

## Rendered Output

![image](https://github.com/quarto-dev/quarto-cli/assets/61118160/f6ed17f7-4bb4-4ac9-b013-3d6a979b33e0)

**Comment by mcanouil** [comment]
I managed to get a really small "working" example.

<table>
<tr><th>Engine</th><th>Quarto document</th><th>HTML</th></tr>
<tr>
<td><code>knitr</code></td>
<td>

````qmd
---
format: html
engine: knitr
---

```{python}
#| layout-ncol: 2
#| label: fig-grouped_signals
#| fig-cap: "Grouped Signals"
#| fig-subcap: 
#|     - "A"
#|     - "B"
#|     - "C"
#|     - "D"
#|     - "E"
#|     - "F"

import matplotlib.pyplot as plt

for i in range(1, 7):
  plt.plot([i])
  plt.show()
```

````

</td>
<td>
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/edfafe55-1285-4ce7-b1e8-aa244f5288cb">
</td>
</tr>
<tr>
<td><code>Jupyter</code></td>
<td>

````qmd
---
format: html
engine: jupyter
---

```{python}
#| layout-ncol: 2
#| label: fig-grouped_signals
#| fig-cap: "Grouped Signals"
#| fig-subcap: 
#|     - "A"
#|     - "B"
#|     - "C"
#|     - "D"
#|     - "E"
#|     - "F"

import matplotlib.pyplot as plt

for i in range(1, 7):
  plt.plot([i])
  plt.show()
```

````

</td>
<td>
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/22e8dc42-1eba-49a4-99d9-32527348402c">
</td>
</tr>
</table>

**Comment by kmink3225** [comment]
Hello, thanks for answering my question. I will make good use of quarto by referring to what you said. Thank you for always working hard to improve quarto :)

**Comment by cderv** [comment]
it seems indeed to be an issue with **knitr**. Possibly with **reticulate** as in the case with **knitr** and Python, I can see 7 plots shown on the HTML. As there is only 6 `fig-subcap`, it seems to lead to some issue. Still trying to identify if this is upstream or in our Quarto R processing though... 🤔 Don't think it is a crossref issue

**Comment by mcanouil** [comment]
There are 7 plots produced in the whole document, but only 6 in the code cell with the weird labelling.

**Comment by cderv** [comment]
Not sure to follow. Copy pasting the **knitr** example gave me 7 plots with an empty one 
![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/2a9fa545-d549-4ea1-8ece-795332fc7cda)

This give me the hint that something is not right with **knitr** + reticulate. Issue is with the internal `options$fig.cur` number that is supposed to select the right option to recycle in the option vector. It does not work correctly which leads to wrong subcap to be selected (because wrong `fig.cur`). 

I think we can reproduce with knitr only - I'll open an issue there.

**Comment by mcanouil** [comment]
Oh, indeed, I did not notice when testing.
The issue seems to be upstream.

````qmd
---
output: html_document
---

```{python}
#| layout-ncol: 2
#| label: fig-grouped_signals
#| fig-cap: "Grouped Signals"
#| fig-subcap: 
#|     - "A"
#|     - "B"

import matplotlib.pyplot as plt

for i in range(1, 3):
    plt.plot([i])
    plt.show()
```

````

<img width="1509" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/d81e8d61-c4a4-4977-b1b4-77393d93e4ec">

**Comment by kmink3225** [comment]
Thank you for the update! I'm glad to hear that the issue has been resolved in reticulate, and it should now work as expected with Quarto.

However, I am currently facing some challenges with installing the reticulate dev version. Once I manage to install it, I will give it a try and confirm whether the problem is indeed resolved for me.

If, despite the fix, I continue to experience any issues or if the problem persists, I will not hesitate to reopen this issue. I genuinely appreciate the efforts put into this fantastic open source, and I am thankful for the support provided.

**Comment by cderv** [comment]
> However, I am currently facing some challenges with installing the reticulate dev version. Once I manage to install it, I will give it a try and confirm whether the problem is indeed resolved for me.

You can try installing from r-universe which builds binary for windows 👍 
````r
install.packages('reticulate', repos = c('https://rstudio.r-universe.dev', 'https://cloud.r-project.org'))
````

**Comment by kmink3225** [comment]
Hello @cderv,

Thank you for your kind explanation.

Previously, I encountered difficulties installing the reticulate package due to conflicts between the R version in Conda and the one on my local PC. Once I resolved these conflicts, I attempted to confirm if the bug had been fixed.

Unfortunately, I have encountered another bug. The code provided works correctly without the subcaption bug:

## Code without Bug

````

```{python}
import matplotlib.pyplot as plt
import numpy as np
```

```{python}
#| layout-ncol: 2
#| label: fig-grouped_signals
#| fig-cap: "Grouped Signals"
#| fig-subcap: 
#|     - "A"
#|     - "B"
#|     - "C"
#|     - "D"
#|     - "E"
#|     - "F"

for i in range(1, 7):
  plt.plot([i])
  plt.show()
```


````

## Code with Bug

 However, when I add an additional plot using another chunk, the bug resurfaces:

````
```{python}
import matplotlib.pyplot as plt
import numpy as np
```

```{python}
#| label: fig-one_plot
#| fig-cap: "one_plot"
#| fig-subcap: 
#|     - "A"
#| eval: true  

plt.plot([0])
plt.show()

```


```{python}
#| layout-ncol: 2
#| label: fig-grouped_signals
#| fig-cap: "Grouped Signals"
#| fig-subcap: 
#|     - "A"
#|     - "B"
#|     - "C"
#|     - "D"
#|     - "E"
#|     - "F"

for i in range(1, 7):
  plt.plot([i])
  plt.show()
```
````

## Output with Bug

![image](https://github.com/quarto-dev/quarto-cli/assets/61118160/3bbc50ed-2e04-4cd2-abb9-bd4119cad50a)
